api/logs/v1/rules_labels_enforcer.go

```diff
@@ -0,0 +1,410 @@
+	return func(res *http.Response) error {
+		tenant, ok := authentication.GetTenant(res.Request.Context())
+		if !ok {
+			return errUnknownTenantKey
+		}
+
+		keys, ok := labelKeys[tenant]
+		if !ok {
+			level.Debug(logger).Log("msg", "Skip applying rule label filters", "tenant", tenant)
+			return nil
+		}
+
+		var (
+			matchers    = extractMatchers(res.Request, keys)
+			contentType = res.Header.Get("Content-Type")
+		)
+
+		matcherStr := fmt.Sprintf("%s", matchers)
+		level.Debug(logger).Log("msg", "filtering using matchers", "tenant", tenant, "matchers", matcherStr)
+
+		body, err := io.ReadAll(res.Body)
+		if err != nil {
+			level.Error(logger).Log("msg", err)
+			return err
+		}
+		res.Body.Close()
+
+		b, err := filterRules(body, contentType, matchers)
+		if err != nil {
+			level.Error(logger).Log("msg", err)
+			return err
+		}
+
+		res.Body = io.NopCloser(bytes.NewReader(b))
+		res.ContentLength = int64(len(b))
+
+		return nil
+	}
+}
+
+func extractMatchers(r *http.Request, l []string) map[string]string {
+	queryParams := r.URL.Query()
+	matchers := map[string]string{}
+	for _, name := range l {
+		value := queryParams.Get(name)
+		if value != "" {
+			matchers[name] = value
+		}
+	}
+
+	return matchers
+}
+
+func filterRules(body []byte, contentType string, matchers map[string]string) ([]byte, error) {
+	switch contentType {
+	case contentTypeApplicationJSON:
+		var res prometheusRulesResponse
+		err := json.Unmarshal(body, &res)
+		if err != nil {
+			return nil, err
+		}
+
+		return json.Marshal(filterPrometheusResponse(res, matchers))
+
+	case contentTypeApplicationYAML:
+		var res lokiRulesResponse
+		if err := yaml.Unmarshal(body, &res); err != nil {
+			return nil, err
+		}
+
+		return yaml.Marshal(filterLokiRules(res, matchers))
+
+	default:
+		return nil, errUnknownRulesContentType
+	}
+}
+
+func filterPrometheusResponse(res prometheusRulesResponse, matchers map[string]string) prometheusRulesResponse {
+	if len(matchers) == 0 {
+		res.Data = rulesData{}
+		return res
+	}
+
+	if len(res.Data.RuleGroups) > 0 {
+		filtered := filterPrometheusRuleGroups(res.Data.RuleGroups, matchers)
+		res.Data = rulesData{RuleGroups: filtered}
+	}
+
+	if len(res.Data.Alerts) > 0 {
+		filtered := filterPrometheusAlerts(res.Data.Alerts, matchers)
+		res.Data = rulesData{Alerts: filtered}
+	}
+
+	return res
+}
+
+func filterPrometheusRuleGroups(groups []*ruleGroup, matchers map[string]string) []*ruleGroup {
+	var filtered []*ruleGroup
+
+	for _, group := range groups {
+		var filteredRules []rule
+
+	rules:
+		for _, rule := range group.Rules {
+			for key, value := range matchers {
+				if !rule.Labels().Has(key) || rule.Labels().Get(key) != value {
+					continue rules
+				}
+			}
+
+			filteredRules = append(filteredRules, rule)
+		}
+
+		if len(filteredRules) > 0 {
+			group.Rules = filteredRules
+			filtered = append(filtered, group)
+		}
+	}
+
+	return filtered
+}
```

Comment:
Here's an alternative way of filtering without repeating the code and using the loop labels.
You will just have to implement the interface for all 3 types.
```suggestion
type labeledRule interface {
	GetLabels() labels.Labels
}

func ruleLabelsMatcheMatchers(rule labeledRule, matchers map[string]string) bool {
	for key, value := range matchers {
		labels := rule.GetLabels().Map()
		val, ok := labels[key]
		if !ok || val != value {
			return false
		}
	}
	return true
}

func filterPrometheusRuleGroups(groups []*ruleGroup, matchers map[string]string) []*ruleGroup {
	var filtered []*ruleGroup

	for _, group := range groups {
		var filteredRules []rule
		for _, rule := range group.Rules {
			if ruleLabelsMatcheMatchers(rule, matchers) {
				filteredRules = append(filteredRules, rule)
			}
		}

		if len(filteredRules) > 0 {
			group.Rules = filteredRules
			filtered = append(filtered, group)
		}
	}

	return filtered
}
```